### ISSUE TITLE
Incorrect Deprecation warning for with_subelements

### ISSUE THREAD

**Issue by hyperized** [ansible/ansible#14701]
##### Issue Type:
- Bug Report
##### Ansible Version:

```
ansible 2.0.1.0
```
##### Environment:

```
  Operating System: Ubuntu 14.04.4 LTS
            Kernel: Linux 3.13.0-77-generic
      Architecture: x86_64
```
##### Summary:

A deprecation warning is issued for 'with_subelements' variables that are not written in a 'future release' notation, incorrectly.
##### Steps To Reproduce:

Original playbook:

```
  with_subelements:
    - users
    - authorized_keys
```

Yields:

```
[DEPRECATION WARNING]: Using bare variables is deprecated. Update your playbooks so that the environment value uses the full variable syntax ('{{users}}'). This feature will be removed in a future release. Deprecation warnings can be disabled by setting
deprecation_warnings=False in ansible.cfg.
```

When the playbook is changed according to the 'future release' notation:

```
  with_subelements:
    - "{{ users }}"
    - "{{ authorized_keys }}"
```

Yields:

```
fatal: [host]: FAILED! => {"failed": true, "msg": "Unexpected failure during module execution.", "stdout": ""}
[DEPRECATION WARNING]: Skipping task due to undefined Error, in the future this
 will be a fatal error.. This feature will be removed in a future release. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```

The latter is understandable when it fails to actually parse the values from with_subelements and it expects strings. Either way, one of the two will have to change.

Please advise on how to proceed to ensure our with_subelements is future proof.

**Comment by bcoca** [comment]
so i can reproduce this if either `users` or `authorized_keys` is undefined .. in which case this is also correct.  Can you give an example in which this is not the case?

**Comment by hyperized** [comment]
I looked through our dataset and I cannot find a single place where 'users' or 'authorized_keys' is undefined, its empty (`[]`) at best. I'll have to construct a testcase though to verify your question.

Looking forward a little, we use the following sections:

```
- name: override users when required
  set_fact:
    users: "{{ users | union(users_additional) }}"
  when: users_additional is defined

- name: update SSH authorized keys
  authorized_key:
    user: "{{ item.0.username }}"
    key: "{{ item.1.key }}"
    state: "{{ item.1.state }}"
  with_subelements:
    - users
    - authorized_keys
  tags: ['users']
  when: item.1.key is defined
```

With the following (anonymized) dataset:

```
users:
  - username: hello1
    name: Hello World
    email: "awesome@email.tld"
    groups: ['tty']
    uid: 1337
    passwd: "$6$..."
    authorized_keys:
      - key: "ssh-rsa AAA..."
        state: present

  - username: hello2
    name: Hello World
    email: "awesome@email.tld"
    groups: ['tty']
    uid: 1338
    passwd: "$6$..."
    authorized_keys: []

users_additional:
  - username: hello3
    name: Hello World
    email: "awesome@email.tld"
    groups: ['tty']
    uid: 1339
    passwd:
    public_key:
      ssh-rsa AAA...
    private_key: |
      MIIEpAIBAAKC...==
    authorized_keys: []

  - username: hello4
    name: Hello World
    email: "awesome@email.tld"
    groups: ['tty']
    uid: 1340
    passwd:
    authorized_keys:
      - key: "ssh-rsa AAA..."
        state: present
```

Could it be that our 'when' condition triggers the deprecation warning?

**Comment by dvianello** [comment]
Have you tried with:

``` yaml
with_subelements:
    - "{{ users }}"
    - authorized_keys
```

Fiddled with the same problem for one hour before realising that the second element should be the name of the key in the users object, not a variable. If you go with `"{{authorized_keys}}"` Ansible will look for a variable named `authorized_keys` that does not exists, hence the error...

**Comment by hyperized** [comment]
That actually seems to do the trick!

I propose the documentation is updated accordingly: https://github.com/ansible/ansible/pull/14747 :).

**Comment by hyperized** [comment]
The error now too makes sense:

```
fatal: [server]: FAILED! => {"failed": true, "msg": "ERROR! could not find 'authorized_keys' key in iterated item '{ yada yada }'"}
```

**Comment by bcoca** [comment]
closing via merged #15065
